### ISSUE TITLE
Cannot use type members of type parameters leaves in method signatures

### ISSUE THREAD

**Issue by kyouko-taiga** [lampepfl/dotty#18898]
## Compiler version

> Scala compiler version 3.3.1 -- Copyright 2002-2023, LAMP/EPFL

## Minimized example

```Scala
import scala.collection.mutable

/// A parser combinator producing instances of `E`.
trait Combinator[E] {
  type Context
  def parse(context: Context): Option[E]
}

/// A parser combinator producing elements by applying `action`.
final case class Apply[C, E](val action: C => Option[E]) extends Combinator[E] {
  type Context = C
  def parse(context: Context): Option[E] = action(context)
}

/// A combinator combining the result of other combinators.
final case class Combine[A, B](
    val first: Combinator[A],
    val second: Combinator[B]
)(using
    first.Context =:= second.Context
) extends Combinator[(A, B)] {
  type Context = first.Context
  def parse(context: Context): Option[(A, B)] = ???
}

@main def hello: Unit = {
  val source = (0 to 10).toList.to(mutable.ListBuffer)
  val n: Combinator[Int] = Apply[mutable.ListBuffer[Int], Int](s => { val h = s.head; s.remove(0); Some(h) })
  val parser = Combine(n, n)
  
  parser.parse(source)
}
```

## Output

```
[error] -- [E007] Type Mismatch Error
[error] 27 |  val r = parser.parse(source)
[error]    |                       ^^^^^^
[error]    |           Found:    (source : scala.collection.mutable.ListBuffer[Int])
[error]    |           Required: parser.Context
```

## Expectation

The compiler should figure out that `parser.Context` is indeed `scala.collection.mutable.ListBuffer[Int]`.

**Comment by odersky** [comment]
This is not a simple big fix. It means we have to go much further into dependent types territory, which is at least a SIP, and probably a whole new research project. See also https://github.com/lampepfl/dotty/issues/3964

**Comment by odersky** [comment]
This is fixed by #19395, if one gives the `modularity` language import.
